### PR TITLE
Added strong damping between the simulated ball and field walls

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

While doing simulated testing, the ball can constantly go out of bounds (or go in the net). Right now these ball-wall collisions are very bouncy, which is both unrealistic and annoying. I've changed it so that these collisions are strongly damped, which has the following benefits:
1. It's more realistic. Walls IRL have relatively little bounce.
2. Related to the above, it makes ball placement much easier if the ball isn't bouncing around for 5 seconds after it goes out of bounds. It will stay closer to the wall like in reality.
3. It's easier to see when the ball went in the net (for a goal)

Note that this only affects collisions between the ball and walls, it does not affect how the ball bounces off robots.

### Testing Done

tested locally

### Resolved Issues

None

### Length Justification

None

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jsn0716 — DISMISSED** [review]
LGTM!

**Review by akhilveeraghanta — COMMENTED** [review]
lgtm pending CI passing

**Review by kelvinkoon — COMMENTED** [review]
LGTM (Pending CI)

**Review by jared-paul — DISMISSED** [review]
LGTM! pending CI

**Comment by MathewMacDougall** [comment]
CI is fixed! :pray: 

(also hooray for memory leak checks :+1: )

File: src/software/simulation/physics/physics_object_user_data.h [hunk 1]
```diff
@@ -24,7 +25,7 @@ struct PhysicsObjectUserData
         : type(type), physics_object(object_ptr)
     {
     }
-    PhysicsObjectUserData() = delete;
+    //    PhysicsObjectUserData() = delete;
```
Comment (jonathanlew) [inline src/software/simulation/physics/physics_object_user_data.h]: commented out code

File: src/software/simulation/physics/simulation_contact_listener.cpp [hunk 2]
```diff
@@ -33,6 +33,17 @@ void SimulationContactListener::BeginContact(b2Contact *contact)
...
+        if (auto ball_field_pair = isBallFieldWallContact(user_data_a, user_data_b))
+        {
+            // If the ball hits a wall it's out of bounds, so we stop it to make it easier
+            // to retrieve for ball placement (otherwise it will bounce around the field)
+            PhysicsBall *ball = ball_field_pair->first;
+            // Apply an impulse to nearly stop the ball. It's preferable if it bounces
+            // away from a wall very slightly for realism, and it should make it easier to
+            // retrieve for ball placement
+            ball->applyImpulse(
+                -ball->momentum().normalize(ball->momentum().length() * 0.95));
```
Comment (jonathanlew) [inline src/software/simulation/physics/simulation_contact_listener.cpp]: something less aggressive feels more realistic, e.g.
```suggestion
                -ball->momentum().normalize(ball->momentum().length() * 0.8));
```

This should be defined as a constant in the header

**Review by jonathanlew — CHANGES_REQUESTED** [review]
cool to see, got a couple comments

File: src/software/simulation/physics/physics_object_user_data.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/software/simulation/physics/physics_object_user_data.h]: whoops, removed.

File: src/software/simulation/physics/simulation_contact_listener.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/simulation/physics/simulation_contact_listener.cpp]: done

**Review by jonathanlew — APPROVED** [review]
LGTM! ⛹️

**Review by xyztiger — APPROVED** [review]

**Pull request merged**